refactor: replace sharp with jimp

LeoMiguelB opened the pull request.

LeoMiguelB committed:
refactor: replace sharp with jimp

removal of sharp to use jimp is done

Completely removed usage of sharp to rely on jimp with uses no native
modules.

rebase changes

LeoMiguelB committed:
refactor: replace sharp with jimp

LeoMiguelB commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ build
 *.log
 .DS_Store
 .vscode
+*.tgz
```
Was building local packages to test. I think it'd be convinient to add this. I can remove if any objections.

LeoMiguelB commented on package.json:
```diff
@@ -87,8 +87,8 @@
     "chalk": "^5.6.0",
     "is-unicode-supported": "^2.1.0",
     "iterm2-version": "^5.0.0",
+    "jimp": "^1.6.0",
```
version of jimp wasn't discussed, let me know if anyone has any objections to using the latest

LeoMiguelB commented on src/components/image/Ascii.tsx:
```diff
@@ -68,20 +65,26 @@ function AsciiImage(props: ImageProps) {
+      image.scaleToFit({ w: width, h: height });
+
+      // TODO: LEO: ask for thoughts on hard coding this, assuming intepreting as jpeg is fine
+      const data = await image.getBuffer("image/jpeg");
```
double commenting here just for double reminder. I'm going to dig deeper into this. Same goes for other files that follow the same pattern.

LeoMiguelB commented on src/types/image.ts:
I saw a couple places that used the same types, hence I thought putting it into it's own type file was better. Let me know if any objections.

LeoMiguelB commented on src/utils/image.ts:
```diff
@@ -1,25 +1,12 @@
+    const img = await Jimp.read(src);
+    return img;
+  } catch (error) {
+    // TODO: LEO: wondering if it's possible to intgrate logger from instagram-cli into here
```
Form what i've seen there's no logger implementation in this codebase. let me know if i'm missed somewhting. Otherise I think this could be something good to add just like in instagram-cli

endernoke commented on src/components/image/ITerm2.tsx:
```diff
@@ -279,16 +275,11 @@ function ITerm2Image(props: ImageProps) {
     "\x1b]1337;File=" +
-    `size=${info.size};` +
+    // optional, only used for showing progress
+    // `size=${size};` +
```
`size` is actually needed here despite the iTerm docs because xterm.js (and perhaps some other terminal emulators) need it, see also https://github.com/xtermjs/xterm.js/issues/5326

endernoke commented on src/index.ts:
```diff
@@ -45,13 +45,14 @@ export {
+// export { default as BrailleImage } from "./components/image/Braille.js";
+// export { default as HalfBlockImage } from "./components/image/HalfBlock.js";
+// export { default as SixelImage } from "./components/image/Sixel.js";
+//
```
I've realized the old exports are unnecessary 😂  Could you just delete these lines thanks

LeoMiguelB replied:
I commented these so I could ignore compiling those other files. I still think we need them for advanced usage though? I've uncommented them for now, let me know if i'm missing something.

endernoke commented on src/components/image/Ascii.tsx:
```diff
@@ -68,20 +65,26 @@ function AsciiImage(props: ImageProps) {
-        .raw()
-        .toBuffer({ resolveWithObject: true });
+      // resized in place
+      image.scaleToFit({ w: width, h: height });
```
Why did you use `scaleToFit` here but `resize` elsewhere?

endernoke reviewed (commented):
Thanks for the work 😄 

When I test it the colors are garbled, I think it's an issue in the `to<ProtocolName>` functions:

<img width="1918" height="1057" alt="Image" src="https://github.com/user-attachments/assets/59b02649-510f-4846-9faa-2b72d7fd3610" />

You could reference a working implementation of reading individual pixel colors from jimp here:
https://github.com/sindresorhus/terminal-image/blob/a2d012f48748d2ef01e4e927125eb4dfc18071c9/index.js#L73

endernoke commented on src/components/image/index.tsx:
```diff
@@ -117,7 +101,7 @@ function Image({
     (currentProtocol: ImageProtocolName): ImageProtocolName => {
       switch (currentProtocol) {
         case "kitty":
-          return "iterm2";
```
This function is supposed to get the next best protocol when the current protocol isn't supported, so kitty --> iterm

LeoMiguelB committed:
refactor: replace sharp with jimp

LeoMiguelB committed:
refactor: replace sharp with jimp

LeoMiguelB committed:
refactor: replace sharp with jimp

LeoMiguelB committed:
fixed braille for jimp implementation

in rgba the maximum instenstiy across all channels is 255, hence we had
to normalize against that by dividing by 255

LeoMiguelB committed:
Fixed reszing of images

Fixed resizing and fitting of images.

LeoMiguelB commented:
Thanks for letting me know. Apologies for inactivity.
